DOCSP-12260 Document JsonObject (#57)
mongodb/docs-java: source/fundamentals/data-formats/document-data-format-extended-json.txt

Read Extended JSON
------------------

The following examples show possible strategies for reading an Extended JSON string 
into Java objects. Click the tab that corresponds to the strategy you would
like to learn more about:

.. tabs::

   .. tab:: Document Class
      :tabid: document-class-read

      You can read an Extended JSON string into Java document objects by calling
      the ``parse()`` static method from either the ``Document`` or ``BsonDocument``
      class, depending on which object type you need. This method parses the Extended
      JSON string in any of the formats and returns an instance of that class
      containing the data.
      
      The following example shows how you can use the ``Document`` class to read
      an example Extended JSON string into a ``Document`` object using the
      ``parse()`` method:
      
      .. code-block:: java
      
         String ejsonStr = "{ \"_id\": { \"$oid\": \"507f1f77bcf86cd799439011\"}," +
                           "\"myNumber\": {\"$numberLong\": \"4794261\" }}}";
      
         Document doc = Document.parse(ejsonStr);
         System.out.println(doc);
      

>       The output of the code above resembles the following:

Reviewer: I know this is inherited, but I would lean toward a more casual tone:
```suggestion
      The output of the code above should look something like this:
```

Author: Thanks for the suggestion @cbush, I agree that a more casual tone is beneficial here. A quick grep search shows that "resembles the following" is used in approximately 9 different files in the java-docs repos. Do you think it is worth replacing all of the "resembles the following"  instances with more casual language, or just this one instance? I will pull @ccho-mongodb into this comment thread as this is getting outside the scope of this PR.

Reviewer 2: @cbush Is this a nitpick? I don't know if there's an issue with the current text.

Reviewer 2: @biniona-mongodb you can certainly change it if you find a wording that conforms much more strongly to the intent or is something you find reference to in the style guide. I think recommended version is ok -- slightly less formal voice than what we've been using in the Java docs.

Author: Changed language throughout java-docs in https://github.com/mongodb/docs-java/pull/57/commits/8fa056f59162a4bb7900e1db00b05c55053390c7 such that "resembles the following" is now different versions of "should look something like this"  when appropriate.

Reviewer 3: @cbush, I provided some feedback on @biniona-mongodb's other [PR]() that seems to conflict with advice here. See [Style Guide](https://docs.mongodb.com/meta/style-guide/writing/use-pronouns-carefully/#this). "this" is too ambiguous without a description. Try "the following output" or "output similar to the following example".  The style guide does note the beginning of a sentence, but it creates the same issues at the beginning or the end of a sentence. From what I understand, it's not a great experience for someone using a screen reader and can create machine translation issues down the line. Using just "the following" creates issues as well.

Reviewer: As I mentioned in Slack, my original critique was on the word _resembles_ in phrases like "resembles the following" when "like this" is much simpler and clearer.

That said, I think having the code example directly following a colon should provide enough context.

                           "\"myNumber\": {\"$numberLong\": \"4794261\" }}}";
      
         JsonReader jsonReader = new JsonReader(ejsonStr);
      
         jsonReader.readStartDocument();
      
         jsonReader.readName("_id");
         ObjectId id = jsonReader.readObjectId();
         jsonReader.readName("myNumber");
         Long myNumber = jsonReader.readInt64();
      
         jsonReader.readEndDocument();
      
         System.out.println(id + " is type: " + id.getClass().getName());
         System.out.println(myNumber + " is type: " + myNumber.getClass().getName());
      
         jsonReader.close();
      
      The output of this code example resembles the following:
      
      .. code-block:: none
         :copyable: False
      
         507f1f77bcf86cd799439011 is type: org.bson.types.ObjectId
         4794261 is type: java.lang.Long
      
      
      For additional information on the classes and methods mentioned on this
      section, see the following API documentation:
      
   .. tab:: JsonObject Class
      :tabid: jsonobject-class-read

      You can read an Extended JSON string into a Java ``JsonObject`` object through the ``JsonObject``

>       class constructor. The :java-docs:`JsonObject <apidocs/bson/org/bson/json/JsonObject.html>`

Reviewer: Why not have this link on the first time you mention JsonObject instead?

Author: Thank for catching this @cbush. While I was writing I thought having the first instance of JsonObject in monospace could be helpful to the reader for placing it in code, but on rereading this section I think it is not particularly helpful and is mainly breaking from conventions elsewhere in the docs. 

I moved the link to the first instance of JsonObject and committed the changes. I will try to place links on the first instance of a term going forward.

      class acts as a wrapper for a JSON string. 
      

>       The following example shows instantiating a ``JsonObject`` with an Extended JSON string:

Reviewer: For crispness, prefer verbs over gerund phrases:
```suggestion
      The following example shows how to instantiate a ``JsonObject`` with an Extended JSON string:
```

Author: Thanks for catching this @cbush, I will keep an eye out for gerund phrases in my writing going forward. Updated with your suggestion.

      .. code-block:: java
      
         String ejsonStr = "{ \"_id\": { \"$oid\": \"507f1f77bcf86cd799439011\"}," +
                           "\"myNumber\": {\"$numberLong\": \"4794261\" }}}";
         JsonObject jsonObj = new JsonObject(ejsonStr);
         System.out.println(jsonObj.getJson());

      The output of this code example resembles the following:
      
      .. code-block:: none
         :copyable: False
      
         { "_id": { "$oid": "507f1f77bcf86cd799439011"},"myNumber": {"$numberLong": "4794261" }}}

      If you want to work primarily with raw JSON strings and avoid intermediate ``Document`` maps in your deployment,
      the ``JsonObject`` class is a more performant alternative than reading JSON to a ``Document`` instance and

>       calling the ``toJson()`` method on that ``Document`` instance. ``JsonObject`` is discussed in more depth

Reviewer: I feel this could be cleaned up a bit. You could also be more direct and prescriptive.

>If you want to work primarily with raw JSON strings and avoid intermediate ``Document`` maps in your deployment, use the ``JsonObject`` class.

Author: Thank you for the note @cbush. I agree that your suggested alternative is cleaner. In my first draft, I think I wanted to try and explain to the reader why the `JsonObject` is a more **performant** alternative, however on rereading with your notes I think it suffices to be prescriptive and just tell the reader to use `JsonObject` in a certain scenario and not go into too much depth as to why that is the case.

I'll try to develop a sense of when and when not to go into a technical explanation when describing a feature/workflow.

      below in the :guilabel:`Write Extended JSON` section under the :guilabel:`JsonObject class` tab.
      
Write Extended JSON
-------------------

The following examples show possible strategies for writing an Extended
JSON string from Java objects. Click the tab that corresponds to the
strategy you would like to learn more about:

.. tabs::

   .. tab:: Document Class
      :tabid: document-class-write

      You can write an Extended JSON string from an instance of ``Document`` or

>       ``BsonDocument`` by calling the ``toJson()`` method, optionally passing it an

Reviewer: The sentence is running a bit long:
```suggestion
      ``BsonDocument`` by calling the ``toJson()`` method. You can pass it an
```

Author: Thanks for the note @cbush , I'll try to keep an eye out for long sentences that can be broken up going forward. 

Updated section with suggested changes.

      Extended JSON string and output it to ``System.out``. We specify the format
      by passing the :java-docs:`outputMode() </apidocs/bson/org/bson/json/JsonWriterSettings.Builder.html#outputMode(org.bson.json.JsonMode)>`
      builder method the ``JsonMode.EXTENDED`` constant:
      
      .. code-block:: java
      
         JsonWriterSettings settings = JsonWriterSettings.builder().outputMode(JsonMode.EXTENDED).build();
      
         try (JsonWriter jsonWriter = new JsonWriter(new BufferedWriter(new OutputStreamWriter(System.out)), settings)) {
             jsonWriter.writeStartDocument();
             jsonWriter.writeObjectId("_id", new ObjectId("507f1f77bcf86cd799439012"));
             jsonWriter.writeInt64("myNumber", 11223344);
             jsonWriter.writeEndDocument();
             jsonWriter.flush();
         }
      
      The output of this code example resembles the following:
      
      .. code-block:: none
         :copyable: false
      
         {"_id": {"$oid": "507f1f77bcf86cd799439012"}, "myNumber": {"$numberLong": "11223344"}}

   .. tab:: JsonObject Class
      :tabid: jsonobject-class-write

>       The ``JsonObject`` class is a wrapper for JSON strings, you can access the

Reviewer: Avoid joining two complete sentences with a comma:
```suggestion
      The ``JsonObject`` class is a wrapper for JSON strings. You can access the
```

Author: Good catch. Updated with suggestion. Will keep an eye out for multiple sentences joined with a comma in my writing going forward.

      string contained in a ``JsonObject`` instance using the ``toJson()`` method.

      The following code snippet shows how to read an extended JSON string from a ``JsonObject``
      using the ``toJson()`` method:

      .. code-block:: java
      
         String ejsonStr = "{ \"_id\": { \"$oid\": \"507f1f77bcf86cd799439011\"}," +
                           "\"myNumber\": {\"$numberLong\": \"4794261\" }}}";
         JsonObject jsonObj = new JsonObject(ejsonStr);
         System.out.println(jsonObj.getJson());

      The output of this code example resembles the following:
      
      .. code-block:: none
         :copyable: False
      
         { "_id": { "$oid": "507f1f77bcf86cd799439011"},"myNumber": {"$numberLong": "4794261" }}}

      The ``JsonObject`` class may not seem very useful at first glance, but its usefulness
      is easier to see when looked at in the context of the
      :java-docs:`default codec registry <apidocs/mongodb-driver-core/com/mongodb/MongoClientSettings.html?is-external=true#getDefaultCodecRegistry()>`
      and the :java-docs:`JsonObjectCodec <apidocs/bson/org/bson/codecs/JsonObjectCodec.html>` class.

>       Below is an example based off of the :doc:`quickstart </fundamentals/data-formats/codecs>`

Reviewer: You can tidy up the verb phrase:
```suggestion
      Below is an example based on the :doc:`quickstart </fundamentals/data-formats/codecs>`
```

Author: Thanks for the note, updated in both places this 'off of' language was used. I agree 'on' is cleaner and easier to understand.

>       Below is an example based off of the :doc:`quickstart </fundamentals/data-formats/codecs>`

Reviewer: The structure of these sentences is a bit awkward. You are trying to say two things:
- The example is based on another example.
- Both this and the other example use the default codec registry and JsonObject to read and write to a MongoDB instance.

For clarity, is it possible to split this into two sentences?

>The following example is based on X. We use the default codec registry to Y:

Author: Thanks for the note @cbush. I agree that this sentence is trying to do two largely independent tasks. I split up the sentence into two sentences as suggested. The first sentence says the example is based off of something. The second sentence highlights what I want the reader to focus on in the example. 

I have updated both instances in the PR where this 'based off of X, look for Y' anti-pattern occurs.

      
         jsonReader.readStartDocument();
      
         jsonReader.readName("_id");
         ObjectId id = jsonReader.readObjectId();
         jsonReader.readName("myNumber");
         Long myNumber = jsonReader.readInt64();
      
         jsonReader.readEndDocument();
      
         System.out.println(id + " is type: " + id.getClass().getName());
         System.out.println(myNumber + " is type: " + myNumber.getClass().getName());
      
         jsonReader.close();
      
      The output of this code example resembles the following:
      
      .. code-block:: none
         :copyable: False
      
         507f1f77bcf86cd799439011 is type: org.bson.types.ObjectId
         4794261 is type: java.lang.Long
      
      
      For additional information on the classes and methods mentioned on this
      section, see the following API documentation:
      
   .. tab:: JsonObject Class
      :tabid: jsonobject-class-read

      You can read an Extended JSON string into a Java
      :java-docs:`JsonObject <apidocs/bson/org/bson/json/JsonObject.html>`
      object through the ``JsonObject`` class constructor. The ``JsonObject``

>       class acts as a wrapper for a JSON string. 

Author: Thanks for the feedback. Moving `JsonObject` documentation to Documents tab resolves this issue.

Takeaways from feedback:

Be clear about actions user has to take to perform a task. Vague language like "through the class constructor" requires that the user fills in the gaps about how you do something through the class constructor. 

In addition, put emphasis on writing succinct paragraphs.

      
      The following example shows how to instantiate a ``JsonObject`` with an Extended JSON string:

      .. code-block:: java
      
         String ejsonStr = "{ \"_id\": { \"$oid\": \"507f1f77bcf86cd799439011\"}," +
                           "\"myNumber\": {\"$numberLong\": \"4794261\" }}}";
         JsonObject jsonObj = new JsonObject(ejsonStr);

>          System.out.println(jsonObj.getJson());

Author: Thanks for the feedback. Added a comment expressing equivalence in new Code snippet in documents section. Line is here: https://github.com/mongodb/docs-java/pull/57/commits/08afb954ea8440e368089d0da237a1f89fe869b9#diff-f63f89b7dd8d7bbec57f0e97ee7158b8a937b8a38f1ce10186213735e45f3f8aR363

      The output of this code example resembles the following:
      
      .. code-block:: none
         :copyable: False
      
         { "_id": { "$oid": "507f1f77bcf86cd799439011"},"myNumber": {"$numberLong": "4794261" }}}

>       If you want to work primarily with raw JSON strings and avoid intermediate ``Document`` maps in your deployment,

Author: Thanks for the feedback. I added a brief explanation of why reader might want to avoid intermediate document map, removed 'deployment' term, and added an introductory sentence at the top in migration of `JsonObject` documentation to Documents tab. https://github.com/mongodb/docs-java/pull/57/commits/67d593bb0df86f39fc0fc0fb09d4fa28420bff96#diff-f63f89b7dd8d7bbec57f0e97ee7158b8a937b8a38f1ce10186213735e45f3f8aR295-R300

      use the ``JsonObject`` class. ``JsonObject`` is discussed in more depth

>       below in the :guilabel:`Write Extended JSON` section under the :guilabel:`JsonObject class` tab.

Author: Resolved in move to documents section.

Takeaways:

Avoid passive language. Try to improve sense of passive language. Avoid 'is discussed'

      
Write Extended JSON
-------------------

The following examples show possible strategies for writing an Extended
JSON string from Java objects. Click the tab that corresponds to the
strategy you would like to learn more about:

.. tabs::

   .. tab:: Document Class
      :tabid: document-class-write

      You can write an Extended JSON string from an instance of ``Document`` or
      ``BsonDocument`` by calling the ``toJson()`` method. You can pass it an
      instance of ``JsonWriterSettings`` to specify the Extended JSON format.
      

>       In this example, we output the Extended JSON in the "Relaxed" format.

Author: Thanks for the feedback. This is resolved in https://github.com/mongodb/docs-java/pull/57/commits/092e992fdbbf5aaf4e6ab8aa15dda37dbd4a7a09

      Extended JSON string and output it to ``System.out``. We specify the format
      by passing the :java-docs:`outputMode() </apidocs/bson/org/bson/json/JsonWriterSettings.Builder.html#outputMode(org.bson.json.JsonMode)>`
      builder method the ``JsonMode.EXTENDED`` constant:
      
      .. code-block:: java
      
         JsonWriterSettings settings = JsonWriterSettings.builder().outputMode(JsonMode.EXTENDED).build();
      
         try (JsonWriter jsonWriter = new JsonWriter(new BufferedWriter(new OutputStreamWriter(System.out)), settings)) {
             jsonWriter.writeStartDocument();
             jsonWriter.writeObjectId("_id", new ObjectId("507f1f77bcf86cd799439012"));
             jsonWriter.writeInt64("myNumber", 11223344);
             jsonWriter.writeEndDocument();
             jsonWriter.flush();
         }
      
      The output of this code example resembles the following:
      
      .. code-block:: none
         :copyable: false
      
         {"_id": {"$oid": "507f1f77bcf86cd799439012"}, "myNumber": {"$numberLong": "11223344"}}

   .. tab:: JsonObject Class
      :tabid: jsonobject-class-write

>       The ``JsonObject`` class is a wrapper for JSON strings. You can access the

Reviewer 2: I think you've already mentioned the same thing in the prior section on reading from JsonObject. You might consider replacing with a more specific description of what it enables you to do in the context of writing data.

Author: Thanks for feedback. This was a consequence of incorrect placement, and was resolved by moving to documents section.

>       string contained in a ``JsonObject`` instance using the ``toJson()`` method.

Author: Response is here: https://github.com/mongodb/docs-java/pull/57#discussion_r581249347

      The following code snippet shows how to read an extended JSON string from a ``JsonObject``

>       using the ``toJson()`` method:

Reviewer 2: Is this a recap of the prior section? Do you think it's necessary in the section on writing?

Author: It was a recap. This issue is addressed in migration of `JsonObject` to Documents section. Recap has to do with incorrect placement of section on first pass.

      .. code-block:: java
      
         String ejsonStr = "{ \"_id\": { \"$oid\": \"507f1f77bcf86cd799439011\"}," +
                           "\"myNumber\": {\"$numberLong\": \"4794261\" }}}";
         JsonObject jsonObj = new JsonObject(ejsonStr);
         System.out.println(jsonObj.getJson());

      The output of this code example resembles the following:
      
      .. code-block:: none
         :copyable: False
      
         { "_id": { "$oid": "507f1f77bcf86cd799439011"},"myNumber": {"$numberLong": "4794261" }}}

      The ``JsonObject`` class may not seem very useful at first glance, but its usefulness

>       is easier to see when looked at in the context of the

Author: Thanks for the feedback. On rereading, I can see how this is inappropriate for our docs writing.

Takeaway from Feedback:

Avoid expressing opinions in documentation. Focus on clear concrete information.

      .. code-block:: java
         
         String uri = "mongodb+srv://<user>:<password>@<cluster-url>?retryWrites=true&w=majority";
         MongoClient mongoClient = MongoClients.create(uri);
         MongoDatabase database = mongoClient.getDatabase("sample_mflix");
         MongoCollection<JsonObject> collection = database.getCollection("movies", JsonObject.class);
         JsonObject doc = collection.find(eq("title", "Back to the Future")).first();
         System.out.println(doc.getJson());
         String ejsonStr = "{ \"_id\": { \"$oid\": \"507f1f77bcf86cd799439011\"}," +
                 "\"myNumber\": {\"$numberLong\": \"4794261\" }}}";
         JsonObject testObj = new JsonObject(ejsonStr);
         //add JsonObject
         InsertOneResult addResult = collection.insertOne(testObj);
         System.out.println("was acknowledged: " + addResult.wasAcknowledged());
         //delete JsonObject
         JsonObject deleteOne = new JsonObject("{ \"_id\": { \"$oid\": \"507f1f77bcf86cd799439011\"}}");
         DeleteResult delResult = collection.deleteOne(deleteOne);
         System.out.println("num deleted: " + delResult.getDeletedCount());

>          mongoClient.close();

Author: Thank you for the feedback. This code-snippet was deleted and this section was moved into the Document section as per our 1 on 1 meeting conversation. Code snippets in Document section should address this critique by following Document section conventions.

      The output of this code example resembles the following:
      
      .. code-block:: none
         :copyable: false
      

>          {"_id": {"$oid": "573a1398f29313caabce9682"}, "plot": "A young man is accidentally sent 30 years into the past in a time-traveling DeLorean invented by his friend, Dr. Emmett Brown, and must make sure his high-school-age parents unite in order to save his own existence.", "genres": ["Adventure", "Comedy", "Sci-Fi"], "runtime": 116, "metacritic": 86, "rated": "PG", "cast": ["Michael J. Fox", "Christopher Lloyd", "Lea Thompson", "Crispin Glover"], "poster": "https://m.media-amazon.com/images/M/MV5BZmU0M2Y1OGUtZjIxNi00ZjBkLTg1MjgtOWIyNThiZWIwYjRiXkEyXkFqcGdeQXVyMTQxNzMzNDI@._V1_SY1000_SX677_AL_.jpg", "title": "Back to the Future", "fullplot": "Marty McFly, a typical American teenager of the Eighties, is accidentally sent back to 1955 in a plutonium-powered DeLorean \"time machine\" invented by slightly mad scientist. During his often hysterical, always amazing trip back in time, Marty must make certain his teenage parents-to-be meet and fall in love - so he can get back to the future.", "languages": ["English"], "released": {"$date": "1985-07-03T00:00:00Z"}, "directors": ["Robert Zemeckis"], "writers": ["Robert Zemeckis", "Bob Gale"], "awards": {"wins": 19, "nominations": 24, "text": "Won 1 Oscar. Another 18 wins & 24 nominations."}, "lastupdated": {"$date": "2015-09-22T11:42:56Z"}, "year": 1985, "imdb": {"rating": 8.5, "votes": 636511, "id": 88763}, "countries": ["USA"], "type": "movie"}

Author: Thanks for the feedback. This comment was addressed in move to Document section by using contrived dataset in existing document examples. 

I will try to be better about using appropriate contrived data for the context of the documentation.

         was acknowledged: true
         num deleted: 1

      The ``JsonObjectCodec`` allows you to use a ``JsonWriterSettings`` object to define

>       a codec between BSON and ``JsonObject`` instances.

Author: Thanks for the feedback. This comment was addressed in move of JsonObject documentation to the Documents section.

Takeaways from Feedback:

Be careful how you talk about codecs. It is incorrect/awkward to describe a codec as 'between' two things.

      Below is a sample ``JsonObjectCodec`` configuration based on the 

>       :doc:`quickstart </fundamentals/data-formats/codecs>`.

Author: Thanks for the feedback. Quickstart references were removed in move to Documents section.

Take aways from feedback:

I will try to not create dependencies between docs sections and the quickstart going forward.

I will try to use title case when making links to other docs in the repository I am working in.

      We define a codec that lets us read and write to a MongoDB instance using

>       Relaxed JSON strings:

Author: Thanks for the feedback. This comment was addressed in move to Documents section. 

Takeaways from Feedback:

Pay attention for chances to combine sentences, but also be careful to have sentences express one idea. Interestingly Chris Bush has opposite feedback about these two sentences https://github.com/mongodb/docs-java/pull/57#discussion_r579517165

      .. code-block:: java
      
         String uri = "mongodb+srv://<user>:<password>@<cluster-url>?retryWrites=true&w=majority";
         CodecRegistry jsonCodec = fromRegistries(

>                          fromCodecs(

Author: Thanks for the note. I will include static imports similar to how they are done in the documents section going forward.

``BsonDocument`` by calling the ``toJson()`` method, optionally passing it an
instance of ``JsonWriterSettings`` to specify the Extended JSON format.

> In this example, we output the Extended JSON in the Relaxed mode format.

Reviewer 2: It's not that clear what the correct term is since both "Relaxed Mode" and "relaxed format" are used in the same bullet item on https://docs.mongodb.com/manual/reference/mongodb-extended-json/ but whichever one you determine is the better one to use, please make sure to stay consistent with the capitalization.

Author: Thanks for the feedback. Based off the [Sever Docs Extended JSON Section](https://docs.mongodb.com/manual/reference/mongodb-extended-json/), it seems like the correct capitalization is "Relaxed mode" (capital "R" lowercase "m" when not used as a title). I changed the other instance of a non-title instance of "Relaxed mode" to match this formatting in https://github.com/mongodb/docs-java/pull/57/commits/9c3c131ad570f67731a635a8969b6636907f3e8a. Let me know your thoughts on this solution.

   System.out.println(doc.toJson(settings)));

The output of this code should look something like this:

.. code-block:: json

   {"_id": "507f1f77bcf86cd799439012", "createdAt": "2020-09-30T21:00:09Z", "myNumber": 4794261}

> Without specifying the converters, the "Relaxed" JSON output should look something like this:

Reviewer 2: Should this be `Relaxed mode JSON output` instead similar to the other updated versions?

Author: Thanks for the note Chris. I think I left that there as the server docs seem to use both, but I think it makes sense for all instances in the document to be the same. Updated here: https://github.com/mongodb/docs-java/pull/57/commits/cb5e031d44b5bb8c4baab131b5e2cf3a03ae33f4
